polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp

```diff
@@ -0,0 +1,296 @@
+      sycl::AliasAnalysis(useRelaxedAliasing));
+
+  // Populate the solver and run the analyses needed by this analysis.
+  solver.load<dataflow::DeadCodeAnalysis>();
+  solver.load<dataflow::SparseConstantPropagation>();
+  solver.load<ReachingDefinitionAnalysis>(*aliasAnalysis);
+
+  if (failed(solver.initializeAndRun(operation))) {
+    operation->emitError("Failed to run required dataflow analyses");
+    return *this;
+  }
+
+  idRangeAnalysis.initialize(useRelaxedAliasing);
+
+  initialized = true;
+
+  return *this;
+}
+
+std::optional<BufferInformation>
+SYCLBufferAnalysis::getBufferInformationFromConstruction(Operation *op,
+                                                         Value operand) {
+  assert(initialized &&
+         "Analysis only available after successful initialization");
+  assert(isa<LLVM::LLVMPointerType>(operand.getType()) &&
+         "Expecting an LLVM pointer");
+
+  const polygeist::ReachingDefinition *reachingDef =
+      solver.lookupState<polygeist::ReachingDefinition>(op);
+  assert(reachingDef && "expected a reaching definition");
+
+  auto mods = reachingDef->getModifiers(operand);
+  if (!mods || mods->empty())
+    return std::nullopt;
+
+  if (!llvm::all_of(*mods,
+                    [&](const Definition &def) { return isConstructor(def); }))
+    return std::nullopt;
+
+  auto pMods = reachingDef->getPotentialModifiers(operand);
+  if (pMods) {
+    if (!llvm::all_of(
+            *pMods, [&](const Definition &def) { return isConstructor(def); }))
+      return std::nullopt;
+  }
+
+  bool first = true;
+  BufferInformation info;
+  for (const Definition &def : *mods) {
+    if (first) {
+      info = getInformation(def);
+      first = false;
+    } else {
+      info = info.join(getInformation(def), *aliasAnalysis);
+      if (!info.hasConstantSize() &&
+          info.isSubBuffer() != SubBufferLattice::YES)
+        // Early return: As soon as joining of the different information has led
+        // to an info with no fixed size and no definitive sub-buffer
+        // information, we can end the processing.
+        return info;
+    }
+  }
+
+  if (pMods) {
+    for (const Definition &def : *pMods) {
+      info = info.join(getInformation(def), *aliasAnalysis);
+      if (!info.hasConstantSize() &&
+          info.isSubBuffer() != SubBufferLattice::YES)
+        // Early return: As soon as joining of the different information has led
+        // to an info with no fixed size and no definitive sub-buffer
+        // information, we can end the processing.
+        return info;
+    }
+  }
+
+  return info;
+}
+
+bool SYCLBufferAnalysis::isConstructor(const Definition &def) {
+  if (!def.isOperation())
+    return false;
+
+  auto constructor = dyn_cast<sycl::SYCLHostConstructorOp>(def.getOperation());
+  if (!constructor)
+    return false;
+
+  return isa<sycl::BufferType>(constructor.getType().getValue());
+}
+
+BufferInformation SYCLBufferAnalysis::getInformation(const Definition &def) {
```

Comment:
Can we rename this to state we're expecting a constructor?

Reply:
This is the private interface, the public interface makes this clear as it's called `getBufferInformationFromConstruction`.